contracts/pcv/balancer/BalancerPCVDepositWeightedPool.sol

```diff
@@ -0,0 +1,326 @@
+        uint256 _maximumSlippageBasisPoints,
+        address _token,
+        IOracle[] memory _tokenOracles
+    ) BalancerPCVDepositBase(_core, _vault, _rewards, _poolId, _maximumSlippageBasisPoints) {
+        // check that we have oracles for all tokens
+        require(poolAssets.length == _tokenOracles.length, "BalancerPCVDepositWeightedPool: wrong number of oracles.");
+
+        tokenOracles = _tokenOracles;
+
+        // set cached values for token addresses & indexes
+        bool tokenFound = false;
+        address _fei = address(fei());
+        for (uint256 i = 0; i < poolAssets.length; i++) {
+            tokenOraclesMapping[IERC20(address(poolAssets[i]))] = _tokenOracles[i];
+            if (address(poolAssets[i]) == _token) {
+                tokenFound = true;
+                tokenIndexInPool = uint8(i);
+                token = IERC20(address(poolAssets[i]));
+            }
+            if (address(poolAssets[i]) == _fei) {
+                feiInPool = true;
+                feiIndexInPool = uint8(i);
+            }
+        }
+        // check that the token is in the pool
+        require(tokenFound, "BalancerPCVDepositWeightedPool: token not in pool.");
+
+        // check that token used for account is not FEI
+        require(_token != _fei, "BalancerPCVDepositWeightedPool: token must not be FEI.");
+    }
+
+    /// @notice sets the oracle for a token in this deposit
+    function setOracle(address _token, address _newOracle) external onlyGovernorOrAdmin {
+        // we must set the oracle for an asset that is in the pool
+        address oldOracle = address(tokenOraclesMapping[IERC20(_token)]);
+        require(oldOracle != address(0), "BalancerPCVDepositWeightedPool: invalid token");
+
+        // set oracle in the map
+        tokenOraclesMapping[IERC20(_token)] = IOracle(_newOracle);
+
+        // emit event
+        emit OracleUpdate(
+            msg.sender,
+            _token,
+            oldOracle,
+            _newOracle
+        );
+    }
+
+    /// @notice returns total balance of PCV in the Deposit, expressed in "token"
+    function balance() public view override returns (uint256) {
+        (, uint256[] memory balances, ) = vault.getPoolTokens(poolId);
+        uint256[] memory underlyingPrices = _readOracles();
+
+        uint256 _balance = balances[tokenIndexInPool];
+        for (uint256 i = 0; i < balances.length; i++) {
+            if (i != feiIndexInPool && i != tokenIndexInPool) {
+                _balance += balances[i] * underlyingPrices[i] / underlyingPrices[tokenIndexInPool];
+            }
+        }
+
+        uint256 _bptSupply = IWeightedPool(poolAddress).totalSupply();
+        uint256 _bptBalance = IWeightedPool(poolAddress).balanceOf(address(this));
+
+        return _balance * _bptBalance / _bptSupply;
+    }
+
+    // @notice returns the manipulation-resistant balance of tokens & FEI held.
+    function resistantBalanceAndFei() public view override returns (
```

Comment:
How do you feel about just reusing Lenses for this instead of duplicating code?

Reply:
why not. We'd have to improve the Lens to handle more than 2 assets and be able to account only the liquidity owned by the PCVDeposit, and not the whole pool content. But it's doable